lib/puppet/util/execution.rb

```diff
@@ -89,6 +89,47 @@ def self.execpipe(command, failonfail = true)
     output
   end
 
+  # Wait for a process to exit or kill it after a given timeout
+  #
+  # This function polls for the exit status of a child process. If the
+  # supplied timeout is exceeded, then the child is terminated.
+  #
+  # @param child_pid [Integer] The PID of the process to monitor.
+  # @param timeout [Integer] The maximum amount of time to wait, in seconds.
+  #
+  # @return [Array<Integer, Process::Status>] A two element array containing
+  #   the PID of the monitored process along with a Process::Status object
+  #   containing its exit status as returned by Process.waitpid2.
+  def self.wait_with_timeout(child_pid, timeout)
+    # TODO: Use a monotonic clock.
+    deadline = Time.now + timeout
+    # The Process module on Windows only supports :KILL, unless a win32
+    # extension gem is used. Even then, the :TERM supplied is equivalent
+    # to :KILL
+    kill_signal = Puppet.features.microsoft_windows? ? :KILL : :TERM
+
+    loop do
+      # Poll status of child process without blocking. Returns nil
+      # if the process has not exited.
+      status = Process.waitpid2(child_pid, Process::WNOHANG)
```

Comment:
The `waitpid2` function seems to work on Windows --- at least, well enough to pass the test I added that exercises it. Also, the `wait_with_timeout` function was added for the POSIX daemons that use fork+exec, so the block of code where `waitpid2` gets run isn't active on Windows:

https://github.com/puppetlabs/puppet/blob/4.10.4/lib/puppet/agent.rb#L65-L77